Mathlib/ModelTheory/Algebra/Field/IsAlgClosed.lean

```diff
@@ -0,0 +1,241 @@
+import Mathlib.ModelTheory.Algebra.Field.CharP
+import Mathlib.ModelTheory.Satisfiability
+
+/-!
+
+# The First Order Theory of Algebraically Closed Fields
+
+This file defines the theory of algebraically closed fields of characteristic `p`, as well
+as proving completeness of the theory and the Lefschetz Principle.
+
+## Main definitions
+
+* `FirstOrder.Language.Theory.ACF p` : the theory of algebraically closed fields of characteristic
+  `p` as a theory over the language of rings.
+* `FirstOrder.Field.ACF_isComplete` : the theory of algebraically closed fields of characteristic
+  `p` is complete whenever `p` is prime or zero.
+* `FirstOrder.Field.ACF_zero_realize_iff_infinite_ACF_prime_realize` : the Lefschetz principle.
+
+## Implementation details
+
+To apply a theorem about the model theory of algebraically closed fields to a specific
+algebraically closed field `K` which does not have a `Language.ring.Structure` instance,
+you must introduce the local instance `compatibleRingOfRing K`. Theorems whose statement requires
+both a `Language.ring.Structure` instance and a `Field` instance will all be stated with the
+assumption `Field K`, `CharP K p`, `IsAlgClosed K` and `CompatibleRing K` and there are instances
+defined saying that these assumptions imply `Theory.field.Model K` and `(Theory.ACF p).Model K`
+
+## References
+
+The first order theory of algebraically closed fields, along with the Lefschetz Principle and
+the Ax-Grothendieck Theorem were first formalized in Lean 3 by Joseph Hua
+[here](https://github.com/Jlh18/ModelTheoryInLean8) with the master's thesis
+[here](https://github.com/Jlh18/ModelTheory8Report)
+
+-/
+
+variable {K : Type*}
+
+namespace FirstOrder
+
+namespace Field
+
+open Ring FreeCommRing BigOperators Polynomial Language
+
+/-- A generic monic polynomial of degree `n` as an element of the
+free commutative ring in `n+1` variables, with a variable for each
+of the `n` non-leading coefficients of the polynomial and one variable (`Fin.last n`)
+for `X`.  -/
+def genericMonicPoly (n : ℕ) : FreeCommRing (Fin (n + 1)) :=
+  of (Fin.last _) ^ n + ∑ i : Fin n, of i.castSucc * of (Fin.last _) ^ (i : ℕ)
+
+theorem lift_genericMonicPoly [CommRing K] [Nontrivial K] {n : ℕ} (v : Fin (n+1) → K) :
+    FreeCommRing.lift v (genericMonicPoly n) =
+    (((monicEquivDegreeLT n).trans (degreeLTEquiv K n).toEquiv).symm (v ∘ Fin.castSucc)).1.eval
+      (v (Fin.last _)) := by
+  simp only [genericMonicPoly, map_add, map_pow, lift_of, map_sum, map_mul, monicEquivDegreeLT,
+    degreeLTEquiv, Equiv.symm_trans_apply, LinearEquiv.coe_toEquiv_symm, EquivLike.coe_coe,
+    LinearEquiv.coe_symm_mk, Function.comp_apply, Equiv.coe_fn_symm_mk, eval_add, eval_pow, eval_X,
+    eval_finset_sum, eval_monomial]
+
+/-- A sentence saying every monic polynomial of degree `n` has a root. -/
+noncomputable def genericMonicPolyHasRoot (n : ℕ) : Language.ring.Sentence :=
+  (∃' ((termOfFreeCommRing (genericMonicPoly n)).relabel Sum.inr =' 0)).alls
```

Comment:
I would have hoped precedence was correct for the following to work
```suggestion
  (∃' (termOfFreeCommRing (genericMonicPoly n)).relabel Sum.inr =' 0).alls
```
I guess this can be fixed in a later PR